Use generics to share code in fake client-go implementations

skitt opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This is the equivalent of #121439 for fake clientsets.

This adds a generic implementation of a fake clientset, and uses it to
replace the template code in generated fake clientsets for the default
methods. The templates are preserved as-is (or as close as they can
be) for use in extensions, whether for resources or subresources.

Fake clientsets with no extensions are reduced to their main getter,
their specific struct, and their constructor. All method
implementations are provided by the generic implementation. The
dedicated struct is preserved to allow extensions and expansions to be
defined where necessary.

Instead of handling the variants (with/without list, apply) with a
complex sequence of if statements, build up an index into an array
containing the various declarations.

Similarly, instead of calling different action constructors for
namespaced and non-namespaced clientsets, assume the current behaviour
of non-namespaced action creation (equivalent to creating a namespaced
action with an empty namespace) and document that assumption in the
action implementation.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Fake clientsets use a common, generic implementation. The corresponding structs are now private, callers must use the corresponding constructors.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

skitt committed:
Generify fake clientsets

This adds a generic implementation of a fake clientset, and uses it to
replace the template code in generated fake clientsets for the default
methods. The templates are preserved as-is (or as close as they can
be) for use in extensions, whether for resources or subresources.

Fake clientsets with no extensions are reduced to their main getter,
their specific struct, and their constructor. All method
implementations are provided by the generic implementation. The
dedicated struct is preserved to allow extensions and expansions to be
defined where necessary.

Instead of handling the variants (with/without list, apply) with a
complex sequence of if statements, build up an index into an array
containing the various declarations.

Similarly, instead of calling different action constructors for
namespaced and non-namespaced clientsets, assume the current behaviour
of non-namespaced action creation (equivalent to creating a namespaced
action with an empty namespace) and document that assumption in the
action implementation.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt committed:
Run codegen

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt commented:
There’s no rush on this, it’s obviously not appropriate for 1.31 at this stage.

/cc @thockin @jpbetz

Jefftree commented:
/assign @thockin @jpbetz 
/triage accepted

skitt commented on staging/src/k8s.io/code-generator/cmd/client-gen/generators/fake/generator_fake_for_type.go:
```diff
@@ -58,40 +59,14 @@ func (g *genFakeForType) Namers(c *generator.Context) namer.NameSystems {
-	}
-	return false
+	imports = g.imports.ImportLines()
+	imports = append(imports, fmt.Sprintf("clientset \"%s\"", g.realClientPackage))
```
This should be reworked to avoid manually adding the import.

skitt commented on staging/src/k8s.io/client-go/gentype/fake.go:
```diff
@@ -0,0 +1,304 @@
+}
+
+// FakeClientWithList represents a fake client with support for lists.
+type FakeClientWithList[T runtime.Object, L runtime.Object] struct {
```
There’s a type mismatch here with respect to the types required for the constructs being faked: these fakes only require `runtime.Object`, whereas the real constructs currently require `runtime.Object` and `MetaObject`. This can be fixed either by merging #126557 or by changing this PR to use `objectWithMeta`.

skitt replied:
In the interests of getting this in without risking breaking third-parties, I’d prefer to leave this as-is, and we can revisit it later (famous last words).

thockin replied:
What's the downside of using `objectWithMeta` ?

skitt replied:
It forces all types to use `meta.Object`, which turned out to be a limitation in 1.31 :wink: (OpenShift had types that weren’t meta objects). However now that we’ve forced everyone to align with that (which is better overall anyway IMO), I suppose we might as well apply it to fakes too, to avoid inconsistencies; I’ve added that as a third commit in the series.

thockin commented:
@skitt Are you hoping to land this in 32?  If so, rebase and I can try to look tomorrow.

skitt commented:
> @skitt Are you hoping to land this in 32? If so, rebase and I can try to look tomorrow.

Might as well, I’ve rebased it and regenerated the fakes.

skitt commented:
The flakiness in https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/126503/pull-kubernetes-verify/1854520541428322304 is surprising: it should always fail (the generated code wouldn’t compile).

thockin reviewed (approved):
Thanks!

/lgtm
/approve

pacoxu commented:
/test pull-kubernetes-integration

skitt committed:
Align fake client-go clients with the main interface

"Real" clients use objectWithMeta to enforce support for meta.Object;
strictly speaking, fakes don't need this, but it's best to align them
with the real clients to ensure that fakes don't end up allowing types
that can't be used with the real clients.

Signed-off-by: Stephen Kitt <skitt@redhat.com>

skitt commented:
@thockin this needed a rebase, so your lgtm was removed. If the [switch to `objectWithMeta`](https://github.com/kubernetes/kubernetes/pull/126503/commits/736e5560ba6b21247c21f8ed12007e1d6d5fec1a) is OK, could you lgtm again?

skitt commented:
/test pull-kubernetes-e2e-kind-ipv6

thockin commented:
/lgtm

skitt commented:
> Do we need an exception request to get this merged?

Apparently not! :partying_face:

sanchezl commented on staging/src/k8s.io/client-go/gentype/fake.go:
```diff
@@ -0,0 +1,304 @@
+
+	obj, err := c.Fake.
+		Invokes(testing.NewGetActionWithOptions(c.resource, c.ns, name, options), emptyResult)
+	if obj == nil {
+		return emptyResult, err
+	}
```
This seems to be a change in behavior from the non-generic code, where this would return `nil, err`.
